[develop] Refactor setup.py to remove use of global variables.

christinaholtNOAA opened the pull request:
## DESCRIPTION OF CHANGES: 
Global variable use has been removed in `setup.py`, and reduced in `generate_FV3LAM_wflow.py`. The use of globals is a carry-over from the bash era of this utility, and does not meet modern coding standards.

This work is one of the preparation steps for providing dictionary configuration objects to an XML generator (still to come).

I will leave a review of the changes made here shortly after I open the PR.

### Type of change
<!-- Please delete options that are not relevant. Add an X to check off a box. -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## TESTS CONDUCTED: 
<!-- Explicitly state what tests were run on these changes, or if any are still pending (for README or other text-only changes, just put "None required"). Make note of the compilers used, the platform/machine, and other relevant details as necessary. For more complicated changes, or those resulting in scientific changes, please be explicit! -->
<!-- Add an X to check off a box. -->

- [x] hera.intel
- [ ] orion.intel
- [ ] cheyenne.intel
- [ ] cheyenne.gnu
- [ ] gaea.intel
- [ ] jet.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [x] fundamental test suite
- [x] comprehensive tests (specify *which* if a subset was used)

## DEPENDENCIES:
None


## CHECKLIST
<!-- Add an X to check off a box. -->
- [ ] My code follows the style guidelines in the Contributor's Guide
- [ ] I have performed a self-review of my own code using the Code Reviewer's Guide
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes do not require updates to the documentation (explain).
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

## LABELS (optional): 
<!-- If you do not have permissions to add labels to your own PR, request that labels be added here. 
Add an X to check off a box. Delete any unnecessary labels. -->
A Code Manager needs to add the following labels to this PR: 
- [ ] Work In Progress
- [ ] bug
- [ ] enhancement
- [ ] documentation
- [ ] release
- [ ] high priority
- [ ] run_ci
- [ ] run_we2e_fundamental_tests
- [ ] run_we2e_comprehensive_tests
- [ ] Needs Cheyenne test 
- [ ] Needs Jet test 
- [ ] Needs Hera test 
- [ ] Needs Orion test 
- [ ] help wanted

christinaholtNOAA committed:
Move config load out into function.

christinaholtNOAA committed:
WIP

christinaholtNOAA committed:
Merge remote-tracking branch 'ufs-community/develop' into refactor_setup

christinaholtNOAA committed:
Run black.

christinaholtNOAA committed:
A tiny attempt at linting.

christinaholtNOAA committed:
WIP

christinaholtNOAA committed:
Merge remote-tracking branch 'ufs-community/develop' into refactor_setup

christinaholtNOAA committed:
Finishing up tests.

christinaholtNOAA committed:
Merge remote-tracking branch 'ufs-community/develop' into refactor_setup

christinaholtNOAA commented on parm/FV3LAM_wflow.xml:
```diff
@@ -362,11 +362,8 @@ MODULES_RUN_TASK_FP script.
-  {%- set fmtstr=" %0"~ndigits_ensmem_names~"d" -%}
-  {{- fmtstr%m -}}
-{%- endfor %} </var>
+    <var name="{{ ensmem_indx_name }}">{% for m in range(1, num_ens_members+1) %}{{ "%03d " % m }}{% endfor %}</var>
```
Members should be named with a consistent set of zeros so it's predictable from experiment to experiment. I set it to 3 here so that we can support the vast majority of use cases.

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.sh:
```diff
@@ -1026,7 +1030,7 @@ model_ver="we2e""
 # Set NCO mode OPSROOT
 #
-OPSROOT=\"${opsroot}\""
+OPSROOT=\"${opsroot:-$OPSROOT}\""
```
Why not use the default when the user does not provide a value?

danielabdi-noaa replied:
Not related to this particular change, but have you tested that specifying OPSROOT from the command line still works. That and 5 other root directories are passed through the environment on WCOSS2

christinaholtNOAA replied:
@danielabdi-noaa Thanks for bringing that to my attention. I did not notice that was a thing, and have most definitely messed it up. 

I will look into that mechanism a bit more. Just to clarify, those env variables are passed at configuration time?

danielabdi-noaa replied:
Yes, they are available during worklfow generation and will override any config settings for OPSROOT, COMROOT etc.

christinaholtNOAA commented on tests/WE2E/test_configs/wflow_features/config.specify_template_filenames.yaml:
```diff
@@ -6,8 +6,8 @@ user:
-  DIAG_TABLE_TMPL_FN: diag_table
-  FIELD_TABLE_TMPL_FN: field_table
+  DIAG_TABLE_TMPL_FN: diag_table.FV3_GFS_v15p2
+  FIELD_TABLE_TMPL_FN: field_table.FV3_GFS_v15p2
```
The setup.py script no longer does this for us, so we need to specify it here in full.

christinaholtNOAA commented on ush/config_defaults.yaml:
```diff
@@ -55,6 +55,19 @@ user:
+  METPLUS_CONF: '{{ [PARMdir, "metplus"]|path_join }}'
+  MET_CONFIG: '{{ [PARMdir, "met"]|path_join }}'
+  UFS_WTHR_MDL_DIR: '{{ user.UFS_WTHR_MDL_DIR }}'
+
```
The added ability to use Jinja2 templates in the config decouples the setup.py script from the configuration for simple substitutions, path joining, math, etc. I think it also improves general readability and usability, too.

danielabdi-noaa replied:
I love that config files have templating option!

christinaholtNOAA commented on ush/config_defaults.yaml:
```diff
@@ -1646,9 +1799,11 @@ task_run_fcst:
-  WRTCMP_write_groups: 1
-  WRTCMP_write_tasks_per_group: 20
+  WRTCMP_write_groups: ""
+  WRTCMP_write_tasks_per_group: ""
```
If we set these values here, and also provide LAYOUT_X and LAYOUT_Y from the user config, PE_MEMBER01 is computed before it should be. I thought it would be relevant to treat them just as all the other WRTCMP variables below.

christinaholtNOAA commented on ush/constants.yaml:
```diff
@@ -45,6 +54,8 @@ constants:
   #-----------------------------------------------------------------------
   #
+  GTYPE: regional
+  TILE_RGNL: 7
```
No reason to set these constants in setup.py while the others are set here.

christinaholtNOAA commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -37,12 +38,12 @@
 
 
-def generate_FV3LAM_wflow(USHdir, logfile: str = "log.generate_FV3LAM_wflow") -> None:
+def generate_FV3LAM_wflow(ushdir, logfile: str = "log.generate_FV3LAM_wflow") -> None:
```
Here, I'm conforming to Python naming conventions for local variables. The final config files and scripts will use the NCO-complant capitalization.

christinaholtNOAA commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -145,47 +152,55 @@ def generate_FV3LAM_wflow(USHdir, logfile: str = "log.generate_FV3LAM_wflow") ->
             )
         except:
-            logging.exception(
+            logging.info(
```
The existing logging.exception did not actually raise the exception when there was a failure to create an XML from the template.

christinaholtNOAA commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -216,6 +236,13 @@ def generate_FV3LAM_wflow(USHdir, logfile: str = "log.generate_FV3LAM_wflow") ->
+    # dictionaries, instead.
+    import_vars(dictionary=flatten_dict(expt_config))
+    export_vars(source_dict=flatten_dict(expt_config))
+
```
Removing globals from the rest of this script will require considerably more effort than I can afford at this point.

christinaholtNOAA commented on ush/link_fix.py:
```diff
@@ -22,11 +23,36 @@
+    """This file defines a function that links fix files to the target
+    directory for a given SRW experiment. Only links files for one group
+    at a time.
+
```
In general for functions, I have passed the parameters needed by the function as arguments and am returning the fields needed by the config to the caller. The caller is responsible for making modifications to the configuration. This is easier to follow in terms of where the configuration settings are being altered and how.

christinaholtNOAA commented on ush/link_fix.py:
```diff
@@ -318,55 +335,37 @@ def link_fix(cfg_d, file_group):
+            # Create links without halo and tile7, and with "tile1"
+            halo_tile = f"{cres}.{field}.tile{tile_rgnl}.halo{nh0}.nc"
+            no_halo_tile = re.sub(f"tile{tile_rgnl}.halo{nh0}", "tile1", halo_tile)
+            create_symlink_to_file(halo_tile, no_halo_tile, True)
```
An attempt to make the logic more readable.

christinaholtNOAA commented on ush/python_utils/config_parser.py:
```diff
@@ -371,16 +453,20 @@ def update_dict(dict_o, dict_t, provide_default=False):
+        elif k in dict_t.keys():
+            if (not provide_default) or (dict_t[k] is None) or \
+                (len(dict_t[k]) == 0) or ("{{" in dict_t[k]):
+                dict_t[k] = v
```
Recursively handles deep dictionaries.

danielabdi-noaa replied:
Nice!

christinaholtNOAA commented on ush/set_ozone_param.py:
```diff
@@ -98,74 +88,37 @@ def set_ozone_param(
-            """,
-            verbose=VERBOSE,
-            dedent_=False,
-        )
```
Updating the config fields has moved to setup.py

christinaholtNOAA commented on ush/set_predef_grid_params.py:
```diff
@@ -10,76 +10,64 @@
-        params_dict["LAYOUT_Y"] = LAYOUT_Y
-    if BLOCKSIZE is not None:
-        params_dict["BLOCKSIZE"] = BLOCKSIZE
-
```
Handled in setup.py

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -1000,637 +981,195 @@ def get_location(xcs, fmt):
-        RUN_TASK_MAKE_GRID = True
-        RUN_TASK_MAKE_OROG = True
-        RUN_TASK_MAKE_SFC_CLIMO = True
-
```
The logic for this has moved up to line 444, and has also been changed to raise an exception to tell the user that there's no way to get the files, instead of assuming this is what should be done.

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -1000,637 +981,195 @@ def get_location(xcs, fmt):
-            "NY": NY,
-            "NHW": NHW,
-            "STRETCH_FAC": STRETCH_FAC,
-        }
```
This use case should never be realized. If it is, there should be an error. That logic is now near line 602.

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -1640,284 +1179,97 @@ def get_location(xcs, fmt):
-    #
-    # -----------------------------------------------------------------------
-    #
-    settings = {
```
All of these were set in the expt_config already.

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -1640,284 +1179,97 @@ def get_location(xcs, fmt):
+    for sect, sect_keys in expt_config.items():
+        for k, v in sect_keys.items():
+            expt_config[sect][k] = str_to_list(v)
+    extend_yaml(expt_config)
```
After all the validation and value-updating, iterate through filling any remaining jinja templates that may exist in the configuration.

christinaholtNOAA committed:
Cleanup mods found during self review.

christinaholtNOAA committed:
Fix ozone test failure.

christinaholtNOAA committed:
Pass unit tests

danielabdi-noaa commented:
@christinaholtNOAA This is great work! Especially, I like the jinja templating of config files which should bring a lot of power and flexibility to how we setup config files.

christinaholtNOAA committed:
Fix NCO test

christinaholtNOAA committed:
Merge remote-tracking branch 'ufs-community/develop' into refactor_setup

MichaelLueken commented on ush/constants.yaml:
```diff
@@ -30,6 +30,15 @@ constants:
+  # Grid type string, set to regional for SRW
+  #
+  # TILE_RGNL: 7
+  # Tile number set ot 7 for a regional grid in SRW
```
`to` rather than `ot`

MichaelLueken commented on ush/set_predef_grid_params.py:
```diff
@@ -10,76 +10,66 @@
-        BLOCKSIZE,
+        USHdir:      path to the SRW ush directory
+        grid_name    str specifying the predefined grid name.
+        quilting:    bool whether quiliting should be used for output
```
`quilting` rather than `quiliting`

MichaelLueken commented on ush/setup.py:
```diff
@@ -423,540 +242,711 @@ def get_location(xcs, fmt):
-    VX_CONFIG_DIR = PARMdir
-    METPLUS_CONF = os.path.join(PARMdir, "metplus")
-    MET_CONFIG = os.path.join(PARMdir, "met")
+    #mkdir_vrfy(f' -p "{expt_basedir}"')
```
Was this line used for debugging?  If so, please remove it, unless it might be used again in the future.

MichaelLueken commented on ush/setup.py:
```diff
@@ -423,540 +242,711 @@ def get_location(xcs, fmt):
+                )
 
-    else:
+    # Make sure RESTART_INTERVAL is set to an integer value
+    restart_interval = fcst_config.get("RESTART_INTERVAL")
+    if not isinstance(restart_interval, int):
+        try:
+            fcst_config["RESTART_INTERVAL"] = int(restart_interval)
+        except ValueError:
+            raise ValueError(
+                f"\nRESTART_INTERVAL = {restart_interval}, must be an integer value\n"
+            )
```
In PR #503, this test has been removed.

@chan-hoo Should this check still exist, or is it safe to entirely remove it following your changes in #503?

chan-hoo replied:
@MichaelLueken, you're correct. This part should be removed!

MichaelLueken reviewed (commented):
@christinaholtNOAA Excellent work!  I have noted a few minor spelling issues, a potential commented out debug line, and checking on whether the RESTART_INTERVAL integer check is required (@chan-hoo removed it with PR #503 that was merged this morning).

When it comes to black, pylint, and other PEP-8 style guides, is consistency important?  I'm asking because I'm seeing a combination of double-quotes and single-quotes within these changes:

```
                f"""
                When performing sub-hourly post (i.e. SUB_HOURLY_POST set to \"TRUE\"),
                DT_SUBHOURLY_POST_MNTS must be set to a value greater than 0; otherwise,
                sub-hourly output is not really being performed:
                  DT_SUBHOURLY_POST_MNTS = \"{DT_SUBHOURLY_POST_MNTS}\"
                Resetting SUB_HOURLY_POST to \"FALSE\".  If you do not want this, you
                must set DT_SUBHOURLY_POST_MNTS to something other than zero."""

```
followed by:

```
                f'''
                When SUB_HOURLY_POST is set to \"TRUE\",
                DT_SUBHOURLY_POST_MNTS must be set to an integer between 1 and 59,
                inclusive but:
                  DT_SUBHOURLY_POST_MNTS = \"{dt_subhourly_post_mnts}\"'''
```
Is this combination okay?  Should we make it consistent throughout the repository or should this be consistent through a given python script?

While testing the Jenkins script, I'm greeted with the following error message:

```
*********************************************************************
FATAL ERROR:
Experiment generation failed. See the error message(s) printed below.
For more detailed information, check the log file from the workflow
generation script: /scratch1/NCEPDEV/stmp2/Michael.Lueken/ufs-srweather-app/ush/log.generate_FV3LAM_wflow
*********************************************************************

Traceback (most recent call last):
  File "/scratch1/NCEPDEV/stmp2/Michael.Lueken/ufs-srweather-app/ush/generate_FV3LAM_wflow.py", line 776, in <module>
    generate_FV3LAM_wflow(USHdir, wflow_logfile)
  File "/scratch1/NCEPDEV/stmp2/Michael.Lueken/ufs-srweather-app/ush/generate_FV3LAM_wflow.py", line 70, in generate_FV3LAM_wflow
    expt_config = setup(ushdir)
  File "/scratch1/NCEPDEV/stmp2/Michael.Lueken/ufs-srweather-app/ush/setup.py", line 1097, in setup
    res_in_fns = link_fix(
  File "/scratch1/NCEPDEV/stmp2/Michael.Lueken/ufs-srweather-app/ush/link_fix.py", line 296, in link_fix
    cd_vrfy(target_dir)
  File "/scratch1/NCEPDEV/stmp2/Michael.Lueken/ufs-srweather-app/ush/python_utils/filesys_cmds_vrfy.py", line 45, in cd_vrfy
    return os.chdir(*args)
FileNotFoundError: [Errno 2] No such file or directory: '/scratch1/NCEPDEV/stmp2/Michael.Lueken/ufs-srweather-app/fix/fix_lam'
```

Interestingly, even though there is a fatal error, the tests still run and ultimately pass.  It is also important to note that this error message isn't present while running the `.cicd/scripts/srw_test.sh` script on develop.  Have you encountered this with your testing as well?

The log files for the NCO mode tests on Hera are also no longer being symlinked into the corresponding expt_dirs/{test}/log location.  Would it be possible to bring this functionality back?

danielabdi-noaa commented on ush/config_defaults.yaml:
```diff
@@ -145,10 +158,11 @@ platform:
-  WFLOW_MOD_FN: ""
-  BUILD_VER_FN: ""
-  RUN_VER_FN: ""
+  LMOD_PATH: ""
```
This is not needed anymore.

danielabdi-noaa commented on ush/setup.py:
```diff
@@ -1921,24 +1282,21 @@ def get_location(xcs, fmt):
                     {k} = {cfg_v[vkey]}"""
             )
 
-    # add LOGDIR and return flat dict
-    cfg_d.update({"LOGDIR": LOGDIR})
-    return cfg_d
```
In NCO mode just tested that this PR is not producing symlinks to log files under EXPTDIR/log properly. I suspect this is because of how LOGDIR is handled. I recall that if LOGDIR ends up being in `var_defns.sh`, it will not work

christinaholtNOAA replied:
@danielabdi-noaa I was just taking a look at how this is handled. Do you think it could be an appropriate solution to set LOGBASEDIR in the config_default.yaml and var_defns.sh to avoid the namespace collision at run time?

danielabdi-noaa replied:
@christinaholtNOAA Yes, i think that will work! The problem is we have two log directories in NCO mode, one under OPSROOT and another in EXPTDIR for symlinks. So if we can use a separate variable for the latter, we can safely store it in var_defns.

danielabdi-noaa commented on tests/WE2E/run_WE2E_tests.sh:
```diff
@@ -797,7 +797,9 @@ Please correct and rerun."
 #
+  save_USHdir=${USHdir}
   source_config ${USHdir}/config_defaults.yaml
+  USHdir=${save_USHdir}
```
Is this needed because USHdir is now in config_defaults.yaml?

christinaholtNOAA replied:
Yes, that's right. I'll add a comment.

danielabdi-noaa commented on ush/config_defaults.yaml:
```diff
@@ -380,6 +394,33 @@ platform:
+  #-----------------------------------------------------------------------
+  #
+  WORKFLOW_ID: ""
+  #
```
This used to be here in config_defaults before, but moved it to setup.py so that the config files contain only variables that are likely to be changed by the user. That is why fixed file mappings were also moved to parm/. I think in this regard, HOMEdir etc could be put in one yaml file and moved to parm/ directory as well. What do you think about this?

christinaholtNOAA replied:
That makes sense to me. Would it be okay to stay here for this PR and I can open an issue about it for future work?

danielabdi-noaa replied:
Sounds good.

danielabdi-noaa commented on ush/setup.py:
```diff
@@ -37,384 +41,199 @@
+    """
+
+    # HOMEdir is the location of the SRW clone, one directory above ush/
+    home_dir = os.path.abspath(os.path.dirname(__file__) + os.sep + os.pardir)
```
Can we use `homedir` instead of `home_dir` just for the sake of rename_model script?

christinaholtNOAA replied:
Yep!

danielabdi-noaa commented on ush/setup.py:
```diff
@@ -423,540 +242,711 @@ def get_location(xcs, fmt):
-              FCST_MODEL = '{FCST_MODEL}'"""
-        )
+    def get_location(xcs, fmt, expt_cfg):
+        ics_lbcs = expt_cfg.get("data", {}).get("ics_lbcs")
```
Good to learn about get() vs [] difference!

danielabdi-noaa commented on ush/config_defaults.yaml:
```diff
@@ -901,7 +1050,7 @@ task_make_grid:
   #-----------------------------------------------------------------------
   # 
-  GRID_DIR: ""
+  GRID_DIR: '{{ [workflow.EXPTDIR, "grid"]|path_join if workflow_switches.RUN_TASK_MAKE_GRID else [platform.DOMAIN_PREGEN_BASEDIR, workflow.PREDEF_GRID_NAME]|path_join}}'
```
I am curious. Does the yaml-xml work needs all variables to be defined in the config_defaults.yaml and not in setup.py?

christinaholtNOAA replied:
It does not. This is me attempting to get all defaults set in the same place. Before this there wasn't a clear distinction on where they were set. Some in setup, others in config_defaults.

A few that I moved to config_defaults.yaml were a bit borderline on where they should be. This was definitely one.

danielabdi-noaa reviewed (commented):
@christinaholtNOAA  Lots of nice work! I have left a few comments.

MichaelLueken commented:
@christinaholtNOAA More rigorous testing has shown that test configurations which cause fatal errors are not being launched.  The tests that have been identified as not running due to fatal errors are:

**On Hera's Intel fundamental tests -**

`pregen_grid_orog_sfc_climo` with fatal error:
`FileNotFoundError: [Errno 2] No such file or directory: '/scratch1/NCEPDEV/stmp2/Michael.Lueken/ufs-srweather-app/fix/fix_lam'`

**No fatal errors were encountered on Hera's GNU fundamental tests.**

**On Gaea's fundamental tests -** 

All HRRR ics/lbcs configurations encounter the following fatal errror:
```
  File "/lustre/f2/pdata/ncep/Michael.Lueken/ufs-srweather-app/ush/setup.py", line 484, in setup
    extrn_mdl_sysbasedir_ics = get_location(
  File "/lustre/f2/pdata/ncep/Michael.Lueken/ufs-srweather-app/ush/setup.py", line 474, in get_location
    v = ics_lbcs[xcs]
KeyError: 'HRRR'
```
`nco_ensemble` with fatal error:
`FileNotFoundError: [Errno 2] No such file or directory: '/lustre/f2/pdata/ncep/Michael.Lueken/ufs-srweather-app/fix/fix_lam'
`

**No fatal errors were encountered on Orion's fundamental tests.**

**On Jet's fundamental tests -**

`grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2` with fatal error:
```
  File "/mnt/lfs4/HFIP/hfv3gfs/Michael.Lueken/ufs-srweather-app/ush/setup.py", line 484, in setup
    extrn_mdl_sysbasedir_ics = get_location(
  File "/mnt/lfs4/HFIP/hfv3gfs/Michael.Lueken/ufs-srweather-app/ush/setup.py", line 474, in get_location
    v = ics_lbcs[xcs]
KeyError: 'GSMGFS'
```

`nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR` with fatal error:
`FileNotFoundError: [Errno 2] No such file or directory: '/mnt/lfs4/HFIP/hfv3gfs/Michael.Lueken/ufs-srweather-app/fix/fix_lam'`

`specify_DOT_OR_USCORE` with fatal error:
```
  File "/mnt/lfs4/HFIP/hfv3gfs/Michael.Lueken/ufs-srweather-app/ush/setup.py", line 484, in setup
    extrn_mdl_sysbasedir_ics = get_location(
  File "/mnt/lfs4/HFIP/hfv3gfs/Michael.Lueken/ufs-srweather-app/ush/setup.py", line 474, in get_location
    v = ics_lbcs[xcs]
KeyError: 'GSMGFS'
```

`specify_RESTART_INTERVAL` with fatal error:
```
Traceback (most recent call last):
  File "/mnt/lfs4/HFIP/hfv3gfs/Michael.Lueken/ufs-srweather-app/ush/setup.py", line 745, in setup
    fcst_config["RESTART_INTERVAL"] = int(restart_interval)
ValueError: invalid literal for int() with base 10: '1 2 5'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/mnt/lfs4/HFIP/hfv3gfs/Michael.Lueken/ufs-srweather-app/ush/generate_FV3LAM_wflow.py", line 776, in <module>
    generate_FV3LAM_wflow(USHdir, wflow_logfile)
  File "/mnt/lfs4/HFIP/hfv3gfs/Michael.Lueken/ufs-srweather-app/ush/generate_FV3LAM_wflow.py", line 70, in generate_FV3LAM_wflow
    expt_config = setup(ushdir)
  File "/mnt/lfs4/HFIP/hfv3gfs/Michael.Lueken/ufs-srweather-app/ush/setup.py", line 747, in setup
    raise ValueError(
ValueError: 
RESTART_INTERVAL = 1 2 5, must be an integer value
```

**On Cheyenne's Intel fundamental tests -**

`grid_RRFS_CONUScompact_13km_ics_HRRR_lbcs_RAP_suite_HRRR` with fatal error:
```
  File "/glade/scratch/mlueken/ufs-srweather-app/ush/setup.py", line 484, in setup
    extrn_mdl_sysbasedir_ics = get_location(
  File "/glade/scratch/mlueken/ufs-srweather-app/ush/setup.py", line 474, in get_location
    v = ics_lbcs[xcs]
KeyError: 'HRRR'
```

`grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_HRRR` with fatal error:
```
  File "/glade/scratch/mlueken/ufs-srweather-app/ush/setup.py", line 484, in setup
    extrn_mdl_sysbasedir_ics = get_location(
  File "/glade/scratch/mlueken/ufs-srweather-app/ush/setup.py", line 474, in get_location
    v = ics_lbcs[xcs]
KeyError: 'NAM'
```

`specify_RESTART_INTERVAL` with fatal error:
```
Traceback (most recent call last):
  File "/glade/scratch/mlueken/ufs-srweather-app/ush/setup.py", line 745, in setup
    fcst_config["RESTART_INTERVAL"] = int(restart_interval)
ValueError: invalid literal for int() with base 10: '1 2 5'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/glade/scratch/mlueken/ufs-srweather-app/ush/generate_FV3LAM_wflow.py", line 776, in <module>
    generate_FV3LAM_wflow(USHdir, wflow_logfile)
  File "/glade/scratch/mlueken/ufs-srweather-app/ush/generate_FV3LAM_wflow.py", line 70, in generate_FV3LAM_wflow
    expt_config = setup(ushdir)
  File "/glade/scratch/mlueken/ufs-srweather-app/ush/setup.py", line 747, in setup
    raise ValueError(
ValueError: 
RESTART_INTERVAL = 1 2 5, must be an integer value
```

`grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_HRRR` with fatal error:
```
  File "/glade/scratch/mlueken/ufs-srweather-app/ush/setup.py", line 484, in setup
    extrn_mdl_sysbasedir_ics = get_location(
  File "/glade/scratch/mlueken/ufs-srweather-app/ush/setup.py", line 474, in get_location
    v = ics_lbcs[xcs]
KeyError: 'HRRR'
```

`grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta` with fatal error:
```
  File "/glade/scratch/mlueken/ufs-srweather-app/ush/setup.py", line 484, in setup
    extrn_mdl_sysbasedir_ics = get_location(
  File "/glade/scratch/mlueken/ufs-srweather-app/ush/setup.py", line 474, in get_location
    v = ics_lbcs[xcs]
KeyError: 'HRRR'
```

`pregen_grid_orog_sfc_clim` with fatal error:
`FileNotFoundError: [Errno 2] No such file or directory: '/glade/scratch/mlueken/ufs-srweather-app/fix/fix_lam'`

**No fatal errors for Cheyenne's GNU fundamental tests.**

christinaholtNOAA commented:
@danielabdi-noaa @MichaelLueken Thank you for the feedback. I have made a couple of changes to address the minor comments (locally) and resolved the conversations to help me keep track of what's left. I will push those changes with some of the larger changes _hopefully_ later today as I work through the longer to-do list.

christinaholtNOAA committed:
Minor changes from review comments.

christinaholtNOAA committed:
Run black on files I modified.

christinaholtNOAA committed:
LOGDIR can't be in var_defns.sh to avoid namespace collision.

christinaholtNOAA committed:
Spell check.

christinaholtNOAA committed:
Fix failures on machines with no data streams.

Also use env values when set for NCO variables.

christinaholtNOAA commented:
@MichaelLueken I believe I've been able to address several of the failures that you shared above. I am having some issues finding an obvious issue leading to the fix_lam failures. Would you mind sharing more of the stack trace for those? Even just a clue about what line/script the failure occurs in would be super helpful. Thanks!

MichaelLueken commented:
@christinaholtNOAA The stack trace for the fix_lam issue on Hera is:

```
  File "/scratch1/NCEPDEV/stmp2/Michael.Lueken/ufs-srweather-app/ush/setup.py", line 1097, in setup
    res_in_fns = link_fix(
  File "/scratch1/NCEPDEV/stmp2/Michael.Lueken/ufs-srweather-app/ush/link_fix.py", line 296, in link_fix
    cd_vrfy(target_dir)
  File "/scratch1/NCEPDEV/stmp2/Michael.Lueken/ufs-srweather-app/ush/python_utils/filesys_cmds_vrfy.py", line 45, in cd_vrfy
    return os.chdir(*args)
FileNotFoundError: [Errno 2] No such file or directory: '/scratch1/NCEPDEV/stmp2/Michael.Lueken/ufs-srweather-app/fix/fix_lam'
```

I'll double check in the morning to make sure this is the case for the rest of the machines too.

MichaelLueken commented:
@christinaholtNOAA I can confirm that the rest of the `fix_lam` fatal errors give the same stack trace as the one from Hera.

christinaholtNOAA committed:
Make directory before linking files there.

christinaholtNOAA committed:
Perhaps better error catching.

Remove print statements from extend_yaml.

danielabdi-noaa reviewed (approved):
@christinaholtNOAA Thanks for addressing my suggestions! Looks good to me.

danielabdi-noaa commented on ush/config_defaults.yaml:
```diff
@@ -1182,7 +1330,8 @@ task_make_orog:
   OMP_STACKSIZE_MAKE_OROG: "2048m"
-  OROG_DIR: ""
+  OROG_DIR: '{{ [workflow.EXPTDIR, "orog"]|path_join }}'
+  TOPO_DIR: ""
```
TOPO_DIR is renamed to FIXorog so can be removed

danielabdi-noaa commented on ush/config_defaults.yaml:
```diff
@@ -1196,7 +1345,8 @@ task_make_sfc_climo:
   OMP_STACKSIZE_MAKE_SFC_CLIMO: "1024m"
-  SFC_CLIMO_DIR: ""
+  SFC_CLIMO_DIR: '{{ [workflow.EXPTDIR, "sfc_climo"]|path_join }}'
+  SFC_CLIMO_INPUT_DIR: ""
```
SFC_CLIMO_INPUT DIR can be removed too since it became FIXsfc

danielabdi-noaa reviewed (commented):
Minor merge issues.

christinaholtNOAA committed:
Removing relics.

christinaholtNOAA commented:
Okay, I was able to successfully complete all the fundamental tests on Hera yesterday after my changes and do not expect that this morning's changes will make a difference. I can kick off the jet tests to see where this PR stands.

MichaelLueken commented:
@christinaholtNOAA Great to hear!  Please try to run them on Jet, however, I think that Jet might be going down shortly for maintenance.

venitahagerty commented:
Machine: hera
Compiler: intel
Job: WE
Repo location: /scratch1/BMC/zrtrr/rrfs_ci/autoci/pr/1143614160/20221213153512/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 10 experiments
If test failed, please make changes and add the following label back:
ci-hera-intel-WE
Experiment Succeeded on hera: community_ensemble_2mems_stoch
Experiment Succeeded on hera: pregen_grid_orog_sfc_climo
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_2017_gfdlmp_regional
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
Experiment Failed on hera: MET_ensemble_verification
2022-12-13 16:28:07 +0000 :: hfe10 :: Task run_pointstatvx_mem001, jobid=38970055, in state DEAD (FAILED), ran for 21.0 seconds, exit status=256, try=1 (of 1)
All experiments completed

MichaelLueken reviewed (approved):
@christinaholtNOAA I have manually submitted the fundamental tests manually on Hera and  Cheyenne.  All tests have successfully ran on these machines (no fatal errors were encountered).  My concerns have also been addressed, so I will now approve this work.  The Jenkins tests will be submitted first thing tomorrow morning, once Jet is back and (hopefully) the Gaea queue has been cleared.

venitahagerty commented:
Machine: jet
Compiler: intel
Job: WE
Repo location: /lfs1/BMC/nrtrr/rrfs_ci/autoci/pr/1143614160/20221214005019/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 10 experiments
If test failed, please make changes and add the following label back:
ci-jet-intel-WE
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
Experiment Succeeded on jet: specify_DT_ATMOS_LAYOUT_XY_BLOCKSIZE
Experiment Succeeded on jet: custom_ESGgrid
Experiment Succeeded on jet: custom_GFDLgrid
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
Experiment Succeeded on jet: nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: specify_RESTART_INTERVAL
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: specify_DOT_OR_USCORE
All experiments completed

MichaelLueken commented:
@christinaholtNOAA While running the standard 9 fundamental tests on Gaea, I encountered a new fatal error -

`nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR` encountered the following fatal error:

```
  File "/lustre/f2/pdata/ncep/Michael.Lueken/ufs-srweather-app/ush/setup.py", line 1099, in setup
    res_in_fns = link_fix(
  File "/lustre/f2/pdata/ncep/Michael.Lueken/ufs-srweather-app/ush/link_fix.py", line 317, in link_fix
    create_symlink_to_file(fp, fn, relative_link_flag)
  File "/lustre/f2/pdata/ncep/Michael.Lueken/ufs-srweather-app/ush/python_utils/create_symlink_to_file.py", line 40, in create_symlink_to_file
    print_err_msg_exit(
  File "/lustre/f2/pdata/ncep/Michael.Lueken/ufs-srweather-app/ush/python_utils/print_msg.py", line 20, in print_err_msg_exit
    traceback.print_stack(file=sys.stderr)
FATAL ERROR: 
Cannot create symlink to specified target file because the latter does
not exist or is not a file:
    target = 'RRFS_CONUScompact_25km/C_mosaic.halo6.nc'
Exiting with nonzero status.
```

I ran a test using develop and the test was submitted without issue.  Please take a look at what might be happening here.  Thanks!

MichaelLueken commented:
Additional details on the Gaea fatal error:

The file in question should be:

`fix/fix_lam/C403_mosaic.halo6.nc`

I'm not sure why it is looking for `C_mosaic.halo6.nc`

Is this due to the following warning message seen?

```
GRID_DIR not specified!
Setting GRID_DIR = RRFS_CONUScompact_25km
```

christinaholtNOAA commented:
I have definitely found room for improvement in terms of logging messages as I try to debug this one.

I am seeing that there is a bit of a gap in the gaea machine file. It does not specify `DOMAIN_PREGEN_BASEDIR` only `TEST_DOMAIN_PREGEN_BASEDIR`. The `run_WE2E_tests.sh` script fills in `GRID_DIR`, `OROG_DIR`, and `SFC_CLIMO_DIR` with a derivative of `TEST_DOMAIN_PREGEN_BASEDIR` when it runs. 

I will update the logic to match this behavior, but does it make sense to set these three in the test script and NOT set `DOMAIN_PREGEN_BASEDIR`?

christinaholtNOAA committed:
Better logging.

Change logic for handling grid_dir, etc. when specified separate from
pregen_basedir.

MichaelLueken commented:
@christinaholtNOAA I'm certainly open to this, but since the machine files are still necessary, and the locations for the `DOMAIN_PREGEN_DIR` are different across the machines, it make sense to keep this in the machine files, at least for now.

christinaholtNOAA commented:
Since I don't have Gaea access today, I tested this test again on hera, but without the `DOMAIN_PREGEN_BASEDIR` defined in the machine file, and it passed. Fingers crossed that means this would work on gaea.

christinaholtNOAA commented:
@MichaelLueken I'm sorry, my suggestion/question wasn't clear!

I definitely think we need to keep the machine files. I also think it's important to have all the machine files that have staged data include full information about that data. For example, as it currently stands, a user running on Gaea could not run their own NCO configuration easily without setting either `DOMAIN_PREGEN_BASEDIR` or all of the `GRID_DIR`, `OROG_DIR`, and `SFC_CLIMO_DIR` variables manually. 

I think that there are 2 ways to approach it (both could be applied):
- Add `DOMAIN_PREGEN_BASEDIR` to the gaea machine file.
- In the test suite, set `DOMAIN_PREGEN_BASEDIR` for the platform being in the user config (prepared by run_WE2E_tests.sh) when `TEST_DOMAIN_PREGEN_BASEDIR` is provided in a machine file.

One down-side...had those two things been done, we wouldn't have had a test that caught that I got rid of the logic for using a user-specified `GRID_DIR` rather than a default PREGEN_BASEDIR.

MichaelLueken commented:
@christinaholtNOAA Ah, I see.  I agree, adding `DOMAIN_PREGEN_BASEDIR` to the gaea machine file and then setting `DOMAIN_PREGEN_BASEDIR` to `TEST_DOMAIN_PREGEN_BASEDIR` if it has been supplied by the machine file sounds great!  Will you be opening a new PR for this, or adding it to this PR?

Also, I have successfully submitted the nco test on gaea and it is has successfully passed.  No more fatal errors!  I'll go ahead and remove the DO_NOT_MERGE label now.  I'm done for the day, so you can go ahead and merge this work if you are ready.

Thanks again for working with me through the issues!

christinaholtNOAA commented:
@MichaelLueken I'll open a 2nd PR for gaea. I'll go ahead and merge this one now.
